src/dbnode/encoding/proto/encoder.go

```diff
@@ -0,0 +1,820 @@
+	// probably need to be able to reset the encoder and also specify the OStream
+	enc.stream.Reset(enc.newBuffer(capacity))
+	enc.m3tszEncoder.Reset(start, capacity)
+	enc.lastEncoded = nil
+	enc.lastEncodedDP = ts.Datapoint{}
+	enc.unmarshaled = nil
+
+	if enc.schema != nil {
+		enc.customFields = resetCustomFields(enc.customFields, enc.schema)
+	}
+
+	enc.hasEncodedFirstSetOfCustomValues = false
+	enc.closed = false
+	enc.numEncoded = 0
+}
+
+func (enc *Encoder) resetSchema(schema *desc.MessageDescriptor) {
+	enc.schema = schema
+	enc.customFields = resetCustomFields(enc.customFields, enc.schema)
+
+	enc.lastEncoded = dynamic.NewMessage(schema)
+	enc.unmarshaled = dynamic.NewMessage(schema)
+}
+
+// Close closes the encoder.
+func (enc *Encoder) Close() {
+	if enc.closed {
+		return
+	}
+
+	enc.closed = true
+	enc.Reset(time.Time{}, 0)
+	enc.stream.Reset(nil)
+	enc.m3tszEncoder.Reset(time.Time{}, 0)
+
+	if pool := enc.opts.EncoderPool(); pool != nil {
+		pool.Put(enc)
+	}
+}
+
+// Discard closes the encoder and transfers ownership of the data stream to
+// the caller.
+func (enc *Encoder) Discard() ts.Segment {
+	segment := enc.discard()
+	// Close the encoder since its no longer needed
+	enc.Close()
+	return segment
+}
+
+// DiscardReset does the same thing as Discard except it also resets the encoder
+// for reuse.
+func (enc *Encoder) DiscardReset(start time.Time, capacity int) ts.Segment {
+	segment := enc.discard()
+	enc.Reset(start, capacity)
+	return segment
+}
+
+func (enc *Encoder) discard() ts.Segment {
+	return enc.segment(false)
+}
+
+// Bytes returns the raw bytes of the underlying data stream. Does not
+// transfer ownership and is generally unsafe.
+func (enc *Encoder) Bytes() ([]byte, error) {
+	if enc.closed {
+		return nil, errEncoderClosed
+	}
+
+	bytes, _ := enc.stream.Rawbytes()
+	return bytes, nil
+}
+
+func (enc *Encoder) encodeCustomValues(m *dynamic.Message) error {
+	for i, customField := range enc.customFields {
+		iVal, err := m.TryGetFieldByNumber(customField.fieldNum)
+		if err != nil {
+			return fmt.Errorf(
+				"%s error trying to get field number: %d",
+				encErrPrefix, customField.fieldNum)
+		}
+
+		customEncoded := true
+		switch {
+		case isCustomFloatEncodedField(customField.fieldType):
+			if err := enc.encodeTSZValue(i, iVal); err != nil {
+				return err
+			}
+		case isCustomIntEncodedField(customField.fieldType):
+			if err := enc.encodeIntValue(i, iVal); err != nil {
+				return err
+			}
+		case customField.fieldType == cBytes:
+			if err := enc.encodeBytesValue(i, iVal); err != nil {
+				return err
+			}
+		default:
+			customEncoded = false
+		}
+
+		if customEncoded {
+			// Remove the field from the message so we don't include it
+			// in the proto marshal.
+			m.ClearFieldByNumber(customField.fieldNum)
+		}
+	}
+	enc.hasEncodedFirstSetOfCustomValues = true
+
+	return nil
+}
+
+func (enc *Encoder) encodeTSZValue(i int, iVal interface{}) error {
```

Comment:
Damn, this will alloc a float64 I assume?

Reply:
its already allocated because I got it as an `interface{}` from the dynamic protobuf library. So luckily (or unluckily depending on how you think about it) this shouldn't cost me anything

Reply:
Definitely unluckily, ugh.  Oh well.  In the future maybe we can add to the dynamic protobuf library something like `Float64ValueForField(index int) (float64, error)` and since we know the types up front we can call that and avoid the `interface{}` casting.

Could you add that maybe to the followup issue?

Reply:
Tracking here: https://github.com/m3db/m3/pull/1426